library/src/scala/annotation/MainAnnotation.scala

```diff
@@ -0,0 +1,119 @@
+package scala.annotation
+
+/** MainAnnotation provides the functionality for a compiler-generated main class.
+ *  It links a compiler-generated main method (call it compiler-main) to a user
+ *  written main method (user-main).
+ *  The protocol of calls from compiler-main is as follows:
+ *
+ *    - create a `command` with the command line arguments,
+ *    - for each parameter of user-main, a call to `command.argGetter`,
+ *      or `command.varargGetter` if is a final varargs parameter,
+ *    - a call to `command.run` with the closure of user-main applied to all arguments.
+ *
+ *  Example:
+ *  ```scala
+ *  /** Sum all the numbers
+ *   *
+ *   *  @param first Fist number to sum
+ *   *  @param rest The rest of the numbers to sum
+ *   */
+ *  @myMain def sum(first: Int, rest: Int*): Int = first + rest.sum
+ *  ```
+ *  generates
+ *  ```scala
+ *  object foo {
+ *    def main(args: Array[String]): Unit = {
+ *      val cmd = new myMain().command(
+ *        info = new CommandInfo(
+ *          name = "foo.main",
+ *          documentation = "Sum all the numbers",
+ *          parameters = Seq(
+ *            new ParameterInfo("first", "scala.Int", hasDefault=false, isVarargs=false, "Fist number to sum"),
+ *            new ParameterInfo("rest", "scala.Int" , hasDefault=false, isVarargs=true, "The rest of the numbers to sum")
+ *          )
+ *        )
+ *        args = args
+ *      )
+ *      val args0 = cmd.argGetter[Int](0, None) // using cmd.Parser[Int]
+ *      val args1 = cmd.varargGetter[Int] // using cmd.Parser[Int]
+ *      cmd.run(() => sum(args0(), args1()*))
+ *    }
+ *  }
+ *  ```
+ *
+ */
+@experimental
+trait MainAnnotation extends StaticAnnotation:
+  import MainAnnotation.{Command, CommandInfo}
+
+  /** A new command with arguments from `args`
+   *
+   *  @param info The information about the command (name, documentation and info about parameters)
+   *  @param args The command line arguments
+   */
+  def command(info: CommandInfo, args: Array[String]): Command[?, ?]
```

Comment:
I believe the result type needs to be defined on the annotation level, not the command level. This way, we force the user to define the command type that can handle the result type known to be used with this particular annotation.

Why do we need the command abstraction anyway? Why can we not just require the user to implement a single `run` method to define the entire annotation? That is, the entire new annotation is fully described by a single function, of a signature of a `def run(program: () => Result, rawArgs: Array[String], argInfos: CommandInfo): Unit`. Do you have any use cases that cannot be covered by an annotation defined by a single function?

Parsing logic of the arguments can be fully abstracted into the compiler's desugaring and depend on the implicit `Parser`s defined with the annotation, following a common-sense logic to handle default parameters and varargs.

Reply:
How would you get the `Parser` instances in that encoding? The point of having the `def argGetter[T]...` is to have the type and be able to summon the `Parser[T]`.

Reply:
That would happen at the same place in desugaring where `argGetter[T]` is called. The difference is that instead of `argGetter[T]`, `summon[Parser[T]]` will be called, and also there will be supporting logic that would handle which arguments from `Array[String]` get processed with which summoned parsers. This logic will also take into account different styles of specifying parameters from the command line in a common-sense way: e.g. with or without dash syntax, with the possibility of params with default values omitted etc.

If the user needs something more complicated than that common sense, they can always fall-back to the good-old `def main`.

Reply:
Currently `Parser` does not define any API. We would not be able to call any method on it to parse the argument.

Reply:
We also do not know which argument to parse as the `Array[String]` may need some preprocessing.

Reply:
I'll investigate this possibility

Reply:
@anatoliykmetyuk I simplified the interface a significant amount thanks to your suggestions. Now we do not have the `Command` class anymore, everything is in `MainAnnotation`. I still keep the `argGetter`/`varargsGetter` as these are necessary to handle errors while parsing. The current design is as expressive as the old one.

Reply:
Looks much better! See my review comments below.

Also, can you elaborate how `argGetter`/`varargGetter` can be used for error handling the way the summoned parser can't? From the MainProxies doc, parsing of arguments:

```scala
   *           val args0: () => S = annotation.argGetter[S](info.parameters(0), cmd(0), None)
```

Do you believe there is something this construct can express that the following can't regarding error handling:

```scala
   *           val args0: () => S = summon[Parser[S]].parse(info.parameters(0), cmd(0), None)
```

Reply:
The `newMain` uses this to try to parse all arguments before failing. In case of a failure, the error is reported and the program continues trying to parse the following arguments. Then the `run` does not evaluate the program if some parsing errors happened.

See example:
* 'argGetter'/'varargsGetter'/'run': https://github.com/lampepfl/dotty/pull/14841/files#diff-73b166621bb3162b59d2e39763a5883b7d44966e7e7959b260417f9a77d2be24R90-R104
* Parsing and error handling: https://github.com/lampepfl/dotty/pull/14841/files#diff-73b166621bb3162b59d2e39763a5883b7d44966e7e7959b260417f9a77d2be24R173-R184